File: node-backend/src/utils/crypto.ts
import crypto from "crypto";
import { env } from "../config/env";

const KEY_LENGTH = 32;

const getKey = () => {
  const key = Buffer.from(env.brokerKeyMaster, "hex");
  if (key.length !== KEY_LENGTH) {
    throw new Error("BROKER_KEY_MASTER_KEY must be 64 hex chars");
  }
  return key;
};

export const encryptSecret = (value: string) => {
  const iv = crypto.randomBytes(12);
  const key = getKey();
  const cipher = crypto.createCipheriv("aes-256-gcm", key, iv);
  const ciphertext = Buffer.concat([cipher.update(value, "utf8"), cipher.final()]);
  const tag = cipher.getAuthTag();
  return Buffer.concat([iv, tag, ciphertext]).toString("base64");
};

export const decryptSecret = (payload: string) => {
  const data = Buffer.from(payload, "base64");
  const iv = data.subarray(0, 12);
  const tag = data.subarray(12, 28);
  const ciphertext = data.subarray(28);
  const key = getKey();
  const decipher = crypto.createDecipheriv("aes-256-gcm", key, iv);
  decipher.setAuthTag(tag);
  return Buffer.concat([decipher.update(ciphertext), decipher.final()]).toString("utf8");
};

export const maskSecret = (value: string, visible = 4) => {
  if (!value) return "";
  const keep = Math.min(visible, value.length);
  return `${"*".repeat(Math.max(value.length - keep, 0))}${value.slice(-keep)}`;
};

File: algo-backend/app/services/crypto.py
import base64
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes

from app.config import settings


def _load_key() -> bytes:
    key_hex = settings.encryption_key
    if not key_hex:
        raise ValueError("ENCRYPTION_KEY is required")
    key = bytes.fromhex(key_hex)
    if len(key) != 32:
        raise ValueError("ENCRYPTION_KEY must be 64 hex chars")
    return key


def decrypt_secret(payload: str) -> str:
    if not payload:
        return ""
    data = base64.b64decode(payload)
    iv = data[:12]
    tag = data[12:28]
    ciphertext = data[28:]
    key = _load_key()
    decryptor = Cipher(algorithms.AES(key), modes.GCM(iv, tag)).decryptor()
    return (decryptor.update(ciphertext) + decryptor.finalize()).decode("utf-8")
